cmd/quadlet/main.go

```diff
@@ -187,63 +188,73 @@ func resolveUnitDirAdminUser() string {
 	return resolvedUnitDirAdminUser
 }
 
-func appendSubPaths(dirs map[string]struct{}, path string, isUserFlag bool, filterPtr func(string, bool) bool) {
+func appendSubPaths(dirs []string, visitedDirs map[string]struct{}, path string, isUserFlag bool, filterPtr func(string, bool) bool) []string {
 	resolvedPath, err := filepath.EvalSymlinks(path)
 	if err != nil {
 		if !errors.Is(err, fs.ErrNotExist) {
 			Debugf("Error occurred resolving path %q: %s", path, err)
 		}
 		// Despite the failure add the path to the list for logging purposes
 		// This is the equivalent of adding the path when info==nil below
-		dirs[path] = struct{}{}
-		return
-	}
-
-	// If the resolvedPath is already in the map no need to read it again
-	if _, already := dirs[resolvedPath]; already {
-		return
+		return addDirToDirs(path, dirs, visitedDirs)
 	}
 
-	// Don't traverse drop-in directories
-	if strings.HasSuffix(resolvedPath, ".d") {
-		return
-	}
-
-	// Check if the directory should be filtered out
-	if filterPtr != nil && !filterPtr(resolvedPath, isUserFlag) {
-		return
-	}
-
-	stat, err := os.Stat(resolvedPath)
-	if err != nil {
-		if !errors.Is(err, fs.ErrNotExist) {
-			Debugf("Error occurred resolving path %q: %s", path, err)
-		}
-		return
-	}
-
-	// Not a directory nothing to add
-	if !stat.IsDir() {
-		return
+	if skipPath(visitedDirs, resolvedPath, isUserFlag, filterPtr) {
+		return dirs
 	}
 
 	// Add the current directory
-	dirs[resolvedPath] = struct{}{}
+	dirs = addDirToDirs(resolvedPath, dirs, visitedDirs)
 
 	// Read the contents of the directory
 	entries, err := os.ReadDir(resolvedPath)
 	if err != nil {
 		if !errors.Is(err, os.ErrNotExist) {
 			Debugf("Error occurred walking sub directories %q: %s", path, err)
 		}
-		return
+		return dirs
 	}
 
 	// Recursively run through the contents of the directory
 	for _, entry := range entries {
 		fullPath := filepath.Join(resolvedPath, entry.Name())
-		appendSubPaths(dirs, fullPath, isUserFlag, filterPtr)
+		dirs = appendSubPaths(dirs, visitedDirs, fullPath, isUserFlag, filterPtr)
+	}
+
+	return dirs
+}
+
+func addDirToDirs(path string, dirs []string, visitedDirs map[string]struct{}) []string {
```

Comment:
maybe it would have been cleaner to define a helper type
```go
type searchPaths struct {
    sorted []string
    // map to store paths so we can quickly check if we saw them already and not loop in case of symlinks
    visitedDirs map[string]struct{}
}
```
And then define a Add() method on it and pass the struct as pointer around.

But given you have written the code already am I also good with the current version, let me know if you like to change it. If not I will merge it

Reply:
Great idea. I'll implement it

Reply:
@Luap99 Done